### ISSUE TITLE
Reader for updating do not use `@JsonCreator` constructor

### ISSUE THREAD

**Issue by pmihalcin** [FasterXML/jackson-databind#1921]
I have `Validity` immutable value object:

```
@Value
public class Validity {
    public static final String VALID_FROM_CANT_BE_NULL = "Valid from can't be null";
    public static final String VALID_TO_CANT_BE_BEFORE_VALID_FROM = "Valid to can't be before valid from";

    private final LocalDate validFrom;
    private final LocalDate validTo;
 
    public Validity(LocalDate validFrom, LocalDate validTo) {
        checkValidity(validFrom, validTo);
 
        this.validFrom = validFrom;
        this.validTo = validTo;
    }
 
    private void checkValidity(LocalDate validFrom, LocalDate validTo) {
        Assert.notNull(validFrom, VALID_FROM_CANT_BE_NULL);
        if (validTo != null) {
            Assert.isTrue(validFrom.isBefore(validTo) || validFrom.isEqual(validTo), VALID_TO_CANT_BE_BEFORE_VALID_FROM);
        }
    }
}
```

and I have jackson customization in place:

```
abstract static class ValidityMixin {
	@JsonCreator
	public ValidityMixin(@JsonProperty(required = true) LocalDate validFrom,
							LocalDate validTo) {
	}
}
```

When I use `mapper.readValue(...)`, Jackson will use constructor to create `Validity` object thanks to `@JsonCreator`

BUT, when I use `mapper.readerForUpdating(...).readValue(...)`, Jackson doesn't use constructor, it uses reflection based setters and I end up with `Validity` object with broken invariant.

Following test written in Groovy and Spock showcases it (it simulates current behaviour, so second expectation should be the same as the first one)

```
def "deserialize Validity with readValue(...) and .readerForUpdating(...).readValue(...)"() {
	when: "I deserialize using readValue(...)"
	mapper.readValue("{\"validFrom\":\"2018-02-01\",\"validTo\":\"2018-01-31\"}", Validity.class)

	then: "I get mapping exception"
	def e = thrown(JsonMappingException)
	e.cause.message == "Valid to can't be before valid from"

	when: "I deserialize using readerForUpdating(...).readvalue(...)"
	Validity validity = new Validity(LocalDate.now(), LocalDate.now().plusDays(1))
	Validity updatedValidity = mapper.readerForUpdating(validity).readValue("{\"validFrom\":\"2018-02-01\",\"validTo\":\"2018-01-31\"}")

	then: "I don't get mapping exception and Jackson don't use constructor annotated with @JsonCreator"

	with(updatedValidity) {
		validFrom == LocalDate.of(2018, FEBRUARY, 1)
		validTo == LocalDate.of(2018, JANUARY, 31)
	}
}
```

Is there a way to keep it consistent in both situations? I think it should be that way

I use Jackson `com.fasterxml.jackson.core:jackson-databind:2.8.10`

**Comment by cowtowncoder** [comment]
If you are to update existing object, how could constructor or factory method be used? That's not the intent. Intent is explicit NOT to create a new value. I am not sure where the expectation comes from, perhaps it needs to be documented better, but for the specific value to update, no Creators can or should be used. Whether such usage should straight on fail with exception or not is a related question however.

**Comment by pmihalcin** [comment]
I understand. Probably better example would be...
Imagine that `Validity` is contained in another class `Account`:

```
public class Account {

    @Getter
    @Setter
    private Validity validity;

    public Specification(SpecificationValidity validity) {
        Assert.notNull(validity, "Validity can't be null");
        this.validity = validity;
    }	
}
```

And when I want to update already existing `validity`, I would issue PATCH request with request body:
```
{
	"validity": {
		"validTo": "2018-02-02",
		"validFrom": "2018-02-03"
	}
}
```

What I would expect is that `@JsonCreator` of `Validity` would be used to create new immutable instance and that instance would be set to `Account` object.
Currently, what Jackson still uses reflection based setters everywhere.

Does this follow up example make sense?

**Comment by cowtowncoder** [comment]
If I understand this example correctly, Creator would indeed be used if updating reader is used. This because update (merge) is shallow, and only root level objects updated; there is no recursive update. Example does not specify creator method however (no `@JsonCreator`) so it's bit difficult to know for sure. 

Recursive merge can be achieved with `@JsonMerge` (and/or with equivalent config overrides setting default merging), in which case getter + setter would be used.

**Comment by pmihalcin** [comment]
I've put together minimal reproducible sample: https://github.com/pmihalcin/jackson-value-object-update
Test: https://github.com/pmihalcin/jackson-value-object-update/blob/master/src/test/groovy/mihalcin/DeserializationSpec.groovy shows what I mean and it fails when you run it. 
`mapper.readValue(...)` https://github.com/pmihalcin/jackson-value-object-update/blob/master/src/test/groovy/mihalcin/DeserializationSpec.groovy#L23 uses constructor to create `Validity` object, but `mapper.readerForUpdating(account).readValue(...)` doesn't even though I think it should.

What do you think?

**Comment by cowtowncoder** [comment]
Ok. One more thing: please do not use Lombok for any tests. I do not know annotations, fields and/or getters/setters it adds by post-processing. I do not use Lombok with Jackson, and no tests to integrate are allowed to use it, simply because it is too much of a black box. Instead please include kinds of annotations it adds.

What should happen here really depend on actual specific annotations, for Java (I assume Groovy translates simply to Java bytecode and is not a factor here).
Since value to modify exists, its properties should be assigned using setters and/or fields accessible. But since this is not deep merge, value of Object-valued properties, such as `Account` will be created from scratch: existing values, if any, are not used.
Perhaps this is the part of processing that is unexpected?

If so, `@JsonMerge` annotation (or, global defaults) may be used to change behavior to try to use merge.

**Comment by pmihalcin** [comment]
I removed Lombok and pushed new version.

> Since value to modify exists, its properties should be assigned using setters and/or fields accessible. But since this is not deep merge, value of Object-valued properties, such as `Account `will be created from scratch: existing values, if any, are not used.

Did you mean 

> value of Object-valued properties, such as `Validity`?

If yes, it is not created from scratch, it is assigned using field access. So,I think it is BUG.

I tried to play with `@JsonMerge`, but the result is the same.

**Comment by pmihalcin** [comment]
@cowtowncoder Do you happen to have any update on the problem I described above? 

https://github.com/pmihalcin/jackson-value-object-update contains sample which reproduces the problem

You mentioned:
> But since this is not deep merge, value of Object-valued properties, such as Account will be created from scratch: existing values, if any, are not used.
This statement doesn't hold in my example. 

If there is anything not understandable, let me know. We can schedule a call and explain my line of thinking more.

**Comment by cowtowncoder** [comment]
@pmihalcin No update: unfortunately I do not think I will have to work on this any time soon.

**Comment by cowtowncoder** [comment]
Ok. Finally had time to go over older issues, and I see the issue (esp. with minor modifications to make sure there are no actual setters, or matching fields).

I think I agree that ideally in such case `@JsonCreator` annotated creator should be used, and merge annotation effectively ignored. However, if setter (or equivalent fields) are available, I think behavior should be to use them since that will do merge as indicated by referencing field...

But I am not sure if there is actually realistic way to actually achieve that; probably not now (2.x) and maybe never. Problem is that figuring out whether to use creator must be done first, without even knowing what data is available. I guess it would be possible to only use creator if no setters were defined but that seems quite fragile.

I did add a failing test just in case I figure out a way that this could be made to work, but for now I have no idea of how that could happen.

**Comment by cowtowncoder** [comment]
No plans to implement; closing as part of bigger issue clean up.

**Comment by cowtowncoder** [comment]
Decided to have another look at targeted fix.
